[WeiHungLoh] iP

WeiHungLoh opened the pull request:
# Duet Chatbot :+1:
A ~smart~ super *duper* smart chatbot with task management capabilities, such as creating To-Do, Deadline and Event tasks. 

## Features
- Add, delete, and list tasks
- Set deadlines
- Able to mark tasks as done

## How it works:
1. Add a task using todo, deadline or event command 
2. Check all your tasks using list
3. Mark tasks as done
4. Delete tasks that you don't need
5. `bye` to end chatbot session

## Code Example
Here's an example of how the chatbot is initialized:
```java
    public Duet(String filePath) {
        ui = new Ui();
        storage = new Storage(filePath);
        try {
            ArrayList<Task> loadedTasks = storage.load();
            tasks = new TaskList(loadedTasks);
        } catch (Exception e) {
            ui.showLoadingError();
            tasks = new TaskList();
        }
    }
```

## Example of task list when "list" is invoked:
- [ ] Study for CS2103T (by: 6 Jan 2025)
- [X] Submit CS2106 Lab 1
- [ ] Finals period (from: 3 Mar 2025 to: 9 Mar 2025)
 
###
Useful Links
[Duet Chatbot Repository](https://github.com/WeiHungLoh/ip)

> Try this bot now!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

WeiHungLoh committed:
Add Duet's greeting message

WeiHungLoh committed:
Add echos commands

WeiHungLoh committed:
Add ability to store text by user

WeiHungLoh committed:
Add mark tasks as done feature

WeiHungLoh committed:
Add Task class

WeiHungLoh committed:
Add ToDo, Event and Deadline classes

WeiHungLoh committed:
Add conditional check for task size > 1

WeiHungLoh committed:
Add conditional check for ToDo task

WeiHungLoh committed:
Add additional space before "("

WeiHungLoh committed:
Add "from" and "to" details

WeiHungLoh committed:
Update javac and java commands

WeiHungLoh committed:
Update input.txt and EXPECTED.txt

WeiHungLoh committed:
Add full stop to follow git convention

WeiHungLoh committed:
Remove Duke.java

WeiHungLoh committed:
Add DuetException class and its subclasses

WeiHungLoh committed:
Add try and catch blocks for exceptions handling

WeiHungLoh committed:
Add Javadoc comments

WeiHungLoh committed:
Add delete task feature

WeiHungLoh committed:
Add extra conditional check for task deletion

WeiHungLoh committed:
Remove empty space

WeiHungLoh committed:
Add save and load features

WeiHungLoh committed:
Add date formatter for Deadline and Event

WeiHungLoh committed:
Merge branch 'branch-Level-8'

WeiHungLoh committed:
Add Ui, Storage, Parser, TaskList classes

WeiHungLoh committed:
Remove unused imports

WeiHungLoh committed:
Divide classes into java packages

WeiHungLoh committed:
Merge commit '81a9c537afe8492675eb253968939364d65db091'

WeiHungLoh committed:
Add DeadlineTest and EventTest classes

WeiHungLoh committed:
Edit build.gradle

WeiHungLoh committed:
Add Javadoc comments

WeiHungLoh committed:
Split import lines and add space

WeiHungLoh committed:
Add find feature

WeiHungLoh committed:
Merge branch 'branch-A-JavaDoc'

WeiHungLoh committed:
Merge branch 'branch-A-CodingStandard'

WeiHungLoh committed:
Merge branch 'branch-Level-9'

WeiHungLoh committed:
Add additional lines

WeiHungLoh committed:
Adhere to checkstyle standards

WeiHungLoh committed:
Merge branch 'branch-A-CheckStyle'

WeiHungLoh committed:
Add GUI to Duet chatbot

WeiHungLoh committed:
Merge branch 'branch-Level-10'

WeiHungLoh committed:
Add welcome message for GUI

Jasonzhou97 commented on src/main/java/duet/Duet.java:
```diff
@@ -0,0 +1,84 @@
+        try {
+            ArrayList<Task> loadedTasks = storage.load();
+            tasks = new TaskList(loadedTasks);
+        } catch (Exception e) {
```
Should this be more specific? eg Filenotfounderror, DuetException etc

Jasonzhou97 commented on src/main/java/duet/storage/Storage.java:
```diff
@@ -0,0 +1,126 @@
+                String task = scan.nextLine();
+                String taskType = task.substring(3, 4);
+
+                if (taskType.equals("T")) {
```
Maybe can create a method in parser for this since the parsing is common?

Jasonzhou97 commented on src/main/java/duet/task/Deadline.java:
```diff
@@ -0,0 +1,45 @@
+     * @throws EmptyInputException If description is empty.
+     * @throws InvalidInputException If by is empty.
+     */
+    public Deadline(String description, String by) throws EmptyInputException, InvalidInputException {
```
Param by can be more specific eg byTime?

WeiHungLoh committed:
Change chatbot pictures

WeiHungLoh committed:
Add assertions to Parser class

WeiHungLoh committed:
Add additional parsing methods and load storage methods.

Parser class only has two main methods for GUI and CLI. Storage class only has one load method.

Parse and load methods do not adhere to SLAP.

Adding more specific methods for parser and load allows us to maintain the highest level of abstraction.

Let's create more specific parser and load methods in Parser and Storage classes.

WeiHungLoh committed:
Adhere to checkstyle test

WeiHungLoh committed:
Replace parameter of loadToDoTask method in Storage class.

loadToDoTask method takes in the wrong input.

Replacing the parameter of loadToDoTask allows us to load tasks properly.

Let's include task description as the input rather than task type.

WeiHungLoh committed:
Revert "Replace parameter of loadToDoTask method in Storage class."

This reverts commit 8e9254e66c85b3b7ec9849d2ed5157dbcd5e78fd.

WeiHungLoh committed:
Replace parameter of loadToDoTask method in Storage class

loadToDoTask method takes in the wrong input.

Replacing the parameter of loadToDoTask allows us to load tasks properly.

Let's include task description as the input rather than task type.

WeiHungLoh committed:
Merge pull request #1 from WeiHungLoh/branch-A-Assertions

Add assertions to Parser class

WeiHungLoh committed:
Change return type of updateCurrentTaskMessage to String

WeiHungLoh committed:
Merge branch 'master' into branch-A-CodeQuality

WeiHungLoh committed:
Merge pull request #3 from WeiHungLoh/branch-A-CodeQuality

Branch a code quality

WeiHungLoh committed:
Merge branch 'master' of https://github.com/WeiHungLoh/ip

WeiHungLoh committed:
Add function to mark tasks as done in bulk

WeiHungLoh committed:
Merge branch 'branch-BCD-Extension'

WeiHungLoh committed:
Add multi-task deletion and improve exception handling

isawangyx commented on src/main/java/duet/exception/DuetException.java:
```diff
@@ -0,0 +1,31 @@
+ */
+public class DuetException extends Exception {
+    /** Description of error message */
+    protected String description;
```
The class defines a protected String description but it is never used. The error message is already stored via super(description).

isawangyx commented on src/main/java/duet/storage/Storage.java:
```diff
@@ -0,0 +1,174 @@
+                String task = scan.nextLine();
+                String taskType = task.substring(3, 4);
+
+                if (taskType.equals("T")) {
+                    loadToDoTask(tasks, task);
+                } else if (taskType.equals("D")) {
+                    loadDeadlineTask(tasks, task);
+                } else if (taskType.equals("E")) {
+                    loadEventTask(tasks, task);
+                } else {
+                    loadTask(tasks, task);
+                }
```
Instead of multiple if-else statements, consider using a switch-case statement for better readability and maintainability.

isawangyx commented on src/main/java/duet/storage/Storage.java:
```diff
@@ -0,0 +1,174 @@
+     */
+    public void loadDeadlineTask(ArrayList<Task> tasks, String task)
+            throws EmptyInputException, InvalidInputException {
+        boolean isDone = task.substring(6, 7).equals("X") ? true : false;
```
The ternary operator is unnecessary as equals("X") already returns a boolean.

isawangyx commented on src/main/java/duet/storage/Storage.java:
```diff
@@ -0,0 +1,174 @@
+        String[] desc = body.split("\\(");
+        String by = desc[1].replace(")", "").replace("by: ", "");
+
+        if (isDone) {
+            tasks.add(new Deadline(desc[0].trim(), by));
+            tasks.get(tasks.size() - 1).markAsDone();
+        } else {
+            tasks.add(new Deadline(desc[0].trim(), by));
+        }
+    }
```
The tasks.add(...) statement is repeated in both branches. Consider merging into a single statement.

isawangyx reviewed (commented):
Overall, can refactor code to reduce redundant code. Good JavaDoc comments.

WeiHungLoh committed:
Add guard clause

WeiHungLoh committed:
Replace else-if conditional statements with switch statements

WeiHungLoh committed:
Add methods to enforce SLAP

WeiHungLoh committed:
Remove addNormalTask() method

WeiHungLoh committed:
Add personality to chatbot

WeiHungLoh committed:
Edit getMultipleDeletedTasks method

WeiHungLoh committed:
Update README.md

WeiHungLoh committed:
Update README.md

WeiHungLoh committed:
Update README.md

WeiHungLoh committed:
Update README.md

WeiHungLoh committed:
Update README.md

WeiHungLoh committed:
Update README.md

WeiHungLoh committed:
Add screenshot of product

WeiHungLoh committed:
Update README.md

WeiHungLoh committed:
Add user guide

WeiHungLoh committed:
Update README.md

WeiHungLoh committed:
Update README.md

WeiHungLoh committed:
Add multiple Parser classes

WeiHungLoh committed:
Merge branch 'master' of https://github.com/WeiHungLoh/ip

WeiHungLoh committed:
Add Javadoc comments
